src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+    int n, char **errstr);
+int ngx_http_lua_ffi_semaphore_wait(ngx_http_request_t *r,
+    ngx_http_lua_semaphore_t *sem, int wait_ms, u_char *errstr, size_t *errlen);
+static void ngx_http_lua_semaphore_cleanup(void *data);
+static void ngx_http_lua_semaphore_handler(ngx_event_t *ev);
+static void ngx_http_lua_semaphore_timeout_handler(ngx_event_t *ev);
+void ngx_http_lua_ffi_semaphore_gc(ngx_http_lua_semaphore_t *sem);
+
+
+enum {
+    SEMAPHORE_WAIT_SUCC = 0,
+    SEMAPHORE_WAIT_TIMEOUT = 1
+};
+
+
+static ngx_http_lua_semaphore_t *
+ngx_http_lua_alloc_semaphore(void)
+{
+    ngx_http_lua_semaphore_t          *sem;
+    ngx_http_lua_semaphore_t          *iter;
+    ngx_http_lua_main_conf_t          *lmcf;
+    ngx_queue_t                       *q;
+    ngx_uint_t                         i, n;
+    ngx_http_lua_semaphore_mm_block_t *block;
+    ngx_http_lua_semaphore_mm_t       *mm;
+
+    if (!ngx_cycle || !ngx_cycle->conf_ctx) {
+        dd("in ngx_http_lua_alloc_semaphore"
+           "ngx_cycle or ngx_cycle->conf_ctx is null");
+
+        return NULL;
+    }
+
+    lmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                               ngx_http_lua_module);
+
+    if (!lmcf) {
+        dd("in ngx_http_lua_alloc_semaphore lmcf is null");
+        return NULL;
+    }
+
+    mm = lmcf->semaphore_mm;
+
+    if (!ngx_queue_empty(&mm->free_queue)) {
+        q = ngx_queue_head(&mm->free_queue);
+        ngx_queue_remove(q);
+
+        sem = ngx_queue_data(q, ngx_http_lua_semaphore_t, chain);
+        sem->block->used++;
+
+        mm->used++;
+
+        return sem;
+    }
+
+    /* free_queue is empty */
+
+    n = sizeof(ngx_http_lua_semaphore_mm_block_t)
+        + mm->num_per_block * sizeof(ngx_http_lua_semaphore_t);
+
+    block = ngx_alloc(n, ngx_cycle->log);
+
+    if (block == NULL) {
+        return NULL;
+    }
+
+    dd("alloc semaphore block:%p", block);
+
+    mm->cur_epoch++;
+    mm->total += mm->num_per_block;
+    mm->used++;
+
+    block->mm = mm;
+    block->epoch = mm->cur_epoch;
+
+    sem = (ngx_http_lua_semaphore_t *) (block + 1);
+    sem->block = block;
+    sem->block->used = 1;
+
+    for (iter = sem + 1, i = 1; i < mm->num_per_block; i++, iter++) {
+        iter->block = block;
+        ngx_queue_insert_tail(&mm->free_queue, &iter->chain);
+    }
+
+    dd("ngx_http_lua_alloc_semaphore sem: %p", sem);
+
+    return sem;
+}
+
+
+void
+ngx_http_lua_cleanup_semaphore_mm(void *data)
+{
+    ngx_http_lua_semaphore_t            *sem;
+    ngx_http_lua_semaphore_t            *iter;
+    ngx_uint_t                           i;
+    ngx_http_lua_main_conf_t            *lmcf;
+    ngx_queue_t                         *q;
+    ngx_http_lua_semaphore_mm_block_t   *block;
+    ngx_http_lua_semaphore_mm_t         *mm;
+
+    lmcf = (ngx_http_lua_main_conf_t*) data;
+    mm = lmcf->semaphore_mm;
+
+    while (!ngx_queue_empty(&mm->free_queue)) {
```

Comment:
It seems to me that this loop never modifies the `mm->free_queue` queue and thus never terminates. Has this been tested?

BTW, the clang static analyzer reports the following warning:

```
src/ngx_http_lua_semaphore.c:138:13: warning: Use of memory after it is freed
        if (block->used == 0) {
            ^~~~~~~~~~~
```